src/coins.cpp

```diff
@@ -1050,20 +1047,69 @@ bool CCoinsViewCache::HaveScRequirements(const CTransaction& tx, int height)
             if (isCeasedAtHeight(scId, height)!= CSidechain::State::ALIVE) {
                 LogPrintf("ERROR: tx[%s] tries to send funds to scId[%s] already ceased at height = %d\n",
                             txHash.ToString(), scId.ToString(), height);
-                return false;
+                return state.Invalid(error("received fwt for ceased sidechain"),
+                             REJECT_INVALID, "sidechain-fwt-ceased-sidechain");
                 }
-        } else {
+        } else
+        {
             if (!Sidechain::hasScCreationOutput(tx, scId)) {
                 LogPrint("sc", "%s():%d - ERROR: tx [%s] tries to send funds to scId[%s] not yet created\n",
                         __func__, __LINE__, txHash.ToString(), scId.ToString() );
-                return false;
+                return state.Invalid(error("scid does not exists"),
+                     REJECT_INVALID, "sidechain-tx-scid");
             }
         }
 
         LogPrint("sc", "%s():%d - OK: tx[%s] is sending [%s] to scId[%s]\n",
             __func__, __LINE__, txHash.ToString(), FormatMoney(ft.nValue), scId.ToString());
     }
 
+    for(size_t idx = 0; idx < tx.GetVBwtRequestOut().size(); ++idx)
+    {
+        const CBwtRequestOut& mbtr = tx.GetVBwtRequestOut().at(idx);
+        const uint256& scId = mbtr.scId;
+        const boost::optional<libzendoomc::ScVk>* pWMbtrVk = nullptr;
+        if (HaveSidechain(scId))
+        {
+            if (isCeasedAtHeight(scId, height)!= CSidechain::State::ALIVE)
+            {
+                LogPrintf("ERROR: tx[%s] contains mainchain bwt request for scId[%s] already ceased at height = %d\n",
+                            txHash.ToString(), scId.ToString(), height);
+                return state.Invalid(error("received mainchain bwt request for ceased sidechain"),
+                             REJECT_INVALID, "sidechain-btr-ceased-sidechain");
+            }
+            pWMbtrVk = &(this->AccessSidechain(scId)->creationData.wMbtrVk);
+        } else
+        {
+            if (!Sidechain::hasScCreationOutput(tx, scId)) {
+                LogPrint("sc", "%s():%d - ERROR: tx [%s] contains mainchain bwt request for scId[%s] not yet created\n",
+                        __func__, __LINE__, txHash.ToString(), scId.ToString() );
+                return state.Invalid(error("scid does not exists"),
+                     REJECT_INVALID, "sidechain-tx-scid");
+            }
+            for(const auto& scCreationOut: tx.GetVscCcOut())
+            	if (scId == scCreationOut.GetScId())
+            	{
+            		pWMbtrVk = &(scCreationOut.wMbtrVk);
+            		break;
+            	}
+        }
+
+        // TODO shall we check if sc balance is not null?
+
+        // Verify mainchain bwt request proof
+        if (!scVerifier.verifyCBwtRequest(mbtr.scId, mbtr.scUtxoId, mbtr.mcDestinationAddress, mbtr.scFee, mbtr.scProof, *pWMbtrVk))
```

Comment:
Previously `HaveScRequirements` method has implemented only simple checks and all the proof related checks were done later.
Maybe, we should check `verifyCBwtRequest` proof later as well?
For CSW I've added `IsTxCswApplicableToState` similar to `IsCertApplicableToState`. Can be useful for BTRs.

Reply:
Now, I'm not sure about the comment. Let's discuss.

Reply:
agreed to leave as is and check later.